serve/mlc_serve/engine/engine_common.py

```diff
@@ -51,6 +51,18 @@ def get_new_request_state(
 
         prompt_token_ids = tokenizer.encode(prompt)
 
+    # TODO: Currently, always create this. But we only need this for the requests with repetition penalty
+    #       Follow-up and optimize when it has been stabilized.
+    # Create prompt mask for repetition penalty
+    tokens=torch.tensor(prompt_token_ids, dtype=torch.long)
+    vocab_size = request.sampling_params.vocab_size
+    bin_counts = torch.zeros((vocab_size + 1,),
+                             dtype=torch.long,
+                             device=tokens.device)
```

Comment:
Can we clarify where this computation runs? What I want to do as a follow-up is clarify where each operation is running and further async/parallelize if we can.

Reply:
I do not have good design now, but plan to think about. It is only the quick fix. I do not think that it is alone place should be fixed, need to think about sampler design (see also comment below)

Reply:
Oh, I was not suggesting we should do it now haha I just wanted to leave a comment where this computation runs. `device=tokens.device` is cpu i assume?

Reply:
I think yes, it is device of prompt_token_ids, I think they are on CPU. Comment has been added